File: runtime-catalog/src/main/java/org/qubership/integration/platform/runtime/catalog/service/extractor/OperationSchemaExtractor.java
package org.qubership.integration.platform.runtime.catalog.service.extractor;

import com.fasterxml.jackson.databind.JsonNode;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.StringUtils;
import org.qubership.integration.platform.parsers.SpecificationParserException;
import org.qubership.integration.platform.parsers.impl.AsyncapiSpecificationParser;
import org.qubership.integration.platform.parsers.impl.GraphqlSpecificationParser;
import org.qubership.integration.platform.parsers.impl.ProtobufSpecificationParser;
import org.qubership.integration.platform.parsers.impl.SwaggerSpecificationParser;
import org.qubership.integration.platform.parsers.model.ParsedOperation;
import org.qubership.integration.platform.runtime.catalog.exception.exceptions.SpecificationImportException;
import org.qubership.integration.platform.runtime.catalog.model.system.OperationProtocol;
import org.qubership.integration.platform.runtime.catalog.persistence.configs.entity.system.SpecificationSource;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import java.util.function.Consumer;
import java.util.stream.Collectors;

/**
 * Rebuilds an operation's request/response schemas and its specification slice on demand from the raw
 * specification source, replacing the schemas import used to materialize into the {@code operations}
 * table. Extraction reuses each protocol's persistence-free {@code parseOperations} core; it never
 * reimplements schema production.
 *
 * <p>Every read parses the source fresh. Reads fetch one operation at a time on small specs, so there
 * is no cache; a caller that needs the whole document at once takes {@link #extractAll} instead of one
 * {@code extract} per operation.
 */
@Slf4j
@Service
public class OperationSchemaExtractor {

    private static final String SPECIFICATION_FILE_PROCESSING_ERROR =
            "An error occurred during parsing specification file";

    private static final String SYNTHETIC_SOURCE_NAME = "source.proto";
    private static final int KEYS_TO_DESCRIBE = 10;
    private static final ExtractedSchemas EMPTY_SCHEMAS = new ExtractedSchemas(null, null, null);

    private final SwaggerSpecificationParser swaggerSpecificationParser;
    private final AsyncapiSpecificationParser asyncapiSpecificationParser;
    private final GraphqlSpecificationParser graphqlSpecificationParser;
    private final ProtobufSpecificationParser protobufSpecificationParser;

    @Autowired
    public OperationSchemaExtractor(
            SwaggerSpecificationParser swaggerSpecificationParser,
            AsyncapiSpecificationParser asyncapiSpecificationParser,
            GraphqlSpecificationParser graphqlSpecificationParser,
            ProtobufSpecificationParser protobufSpecificationParser) {
        this.swaggerSpecificationParser = swaggerSpecificationParser;
        this.asyncapiSpecificationParser = asyncapiSpecificationParser;
        this.graphqlSpecificationParser = graphqlSpecificationParser;
        this.protobufSpecificationParser = protobufSpecificationParser;
    }

    /**
     * The extractor output behind the stable {@code OperationInfo} contract. {@code requestSchema} and
     * {@code responseSchemas} are {@code null} for protocols that carry no schemas, such as WSDL and
     * GraphQL.
     */
    public record ExtractedSchemas(
            JsonNode specification,
            Map<String, JsonNode> requestSchema,
            Map<String, JsonNode> responseSchemas
    ) {
    }

    /**
     * Key of an operation inside one parsed document: exact path plus case-insensitive method, the pair
     * {@code matchOperation} compares. Build it through {@link #of} so both sides normalize alike.
     */
    public record OperationKey(String path, String method) {

        public static OperationKey of(String path, String method) {
            return new OperationKey(path, method == null ? null : method.toUpperCase(Locale.ROOT));
        }
    }

    /**
     * Renders keys for a log line, capped so one bad document cannot flood the log. Callers that report a key miss
     * share this, so the same miss reads the same way on the import side and on the export side.
     */
    public static String describeKeys(List<OperationKey> keys) {
        String listed = keys.stream()
                .limit(KEYS_TO_DESCRIBE)
                .map(key -> key.method() + " " + key.path())
                .collect(Collectors.joining(", "));
        return keys.size() > KEYS_TO_DESCRIBE
                ? listed + ", and " + (keys.size() - KEYS_TO_DESCRIBE) + " more"
                : listed;
    }

    /**
     * Persistence-free, single-source seam for tests. Wraps the raw source as one
     * {@link SpecificationSource} and delegates to the multi-source method, so the parity gate can drive
     * extraction with corpus fixtures that never touch the database. Production always calls the
     * {@link SpecificationSource}-list overload.
     */
    ExtractedSchemas extract(String rawSource, OperationProtocol protocol, String path, String method) {
        requireProtocol(protocol);
        if (hasNoSchemaExtraction(protocol)) {
            return EMPTY_SCHEMAS;
        }
        SpecificationSource source = SpecificationSource.builder()
                .name(SYNTHETIC_SOURCE_NAME)
                .isMainSource(true)
                .source(rawSource)
                .build();
        return extract(List.of(source), protocol, path, method);
    }

    /**
     * Production seam. Takes every {@link SpecificationSource} the model owns: protobuf needs all of them
     * (a {@code service.proto} can reference a message declared in a second file), while other protocols
     * parse only the main source.
     */
    public ExtractedSchemas extract(
            List<SpecificationSource> sources, OperationProtocol protocol, String path, String method) {
        requireProtocol(protocol);
        if (hasNoSchemaExtraction(protocol) || lacksParseableSource(sources, protocol)) {
            return EMPTY_SCHEMAS;
        }
        List<ParsedOperation> operations = parseDocument(sources, protocol, true);
        return toExtractedSchemas(matchOperation(operations, path, method));
    }

    /**
     * Bulk seam for callers that need every operation of one model, such as the legacy export. The document
     * is parsed once, whereas {@link #extract} reparses it per call and so costs one parse per operation.
     *
     * <p>{@code withSchemas} is the cost switch: the {@code specification} slice is built either way, while
     * request and response schemas inline every referenced component into every operation and every response
     * code. A caller that reads only {@code specification} passes {@code false}.
     *
     * <p>A protocol with no schemas and a source with nothing to parse yield an empty map, and a key claimed
     * by two operations drops out of it — the bulk counterpart of the per-operation throw on an ambiguous
     * match. A parse failure still surfaces, so the caller decides how to degrade.
     */
    public Map<OperationKey, ExtractedSchemas> extractAll(
            List<SpecificationSource> sources, OperationProtocol protocol, boolean withSchemas) {
        requireProtocol(protocol);
        if (hasNoSchemaExtraction(protocol) || lacksParseableSource(sources, protocol)) {
            return Map.of();
        }
        Map<OperationKey, ExtractedSchemas> schemasByOperation = new HashMap<>();
        Set<OperationKey> ambiguousKeys = new HashSet<>();
        for (ParsedOperation operation : parseDocument(sources, protocol, withSchemas)) {
            OperationKey key = OperationKey.of(operation.getPath(), operation.getMethod());
            if (schemasByOperation.put(key, toExtractedSchemas(operation)) != null) {
                ambiguousKeys.add(key);
            }
        }
        for (OperationKey key : ambiguousKeys) {
            schemasByOperation.remove(key);
            log.warn("Ambiguous operation for path='{}', method='{}': its schemas are not extracted",
                    key.path(), key.method());
        }
        return schemasByOperation;
    }

    // Parses on the request thread, inside the caller's read-only transaction, with no size cap: every parser core
    // wraps its failures in SpecificationImportException, so a bad source costs one parse and degrades to null schemas.
    private List<ParsedOperation> parseDocument(
            List<SpecificationSource> sources, OperationProtocol protocol, boolean withSchemas) {
        List<org.qubership.integration.platform.parsers.SpecificationSource> librarySources = toLibrarySources(sources);
        Consumer<String> silent = message -> { };
        try {
            return parseWithLibrary(protocol, librarySources, silent);
        } catch (SpecificationParserException e) {
            // The read path answers with the catalog's import-facing exception, the same one the import path
            // surfaces, so a caller has one type to degrade on rather than two.
            throw new SpecificationImportException(SPECIFICATION_FILE_PROCESSING_ERROR, e);
        }
    }

    private List<ParsedOperation> parseWithLibrary(
            OperationProtocol protocol,
            List<org.qubership.integration.platform.parsers.SpecificationSource> librarySources,
            Consumer<String> silent) {
        return switch (protocol) {
            case HTTP -> swaggerSpecificationParser.parseSpecification(null, librarySources, silent).getOperations();
            case AMQP, KAFKA -> asyncapiSpecificationParser.parseSpecification(null, librarySources, silent).getOperations();
            case GRAPHQL -> graphqlSpecificationParser.parseSpecification(null, librarySources, silent).getOperations();
            case GRPC -> protobufSpecificationParser.parseSpecification(null, librarySources, silent).getOperations();
            default -> List.of(); // SOAP/METAMODEL never reach here: filtered out by hasNoSchemaExtraction above
        };
    }

    // The library parsers read their own source type and always build schemas; `withSchemas` no longer changes
    // the parse, only whether the caller keeps what came back.
    private static List<org.qubership.integration.platform.parsers.SpecificationSource> toLibrarySources(
            List<SpecificationSource> sources) {
        return sources == null ? List.of() : sources.stream()
                .map(source -> new org.qubership.integration.platform.parsers.SpecificationSource(
                        source.getName(), source.getSource(), source.isMainSource()))
                .toList();
    }

    private static String mainSourceText(List<SpecificationSource> sources) {
        if (sources == null) {
            return null;
        }
        return sources.stream()
                .filter(SpecificationSource::isMainSource)
                .findFirst()
                .or(() -> sources.stream().findFirst())
                .map(SpecificationSource::getSource)
                .orElse(null);
    }

    // A blank main source (no content to parse) degrades to EMPTY_SCHEMAS up front instead of handing a
    // null/blank string to a parser. Protobuf is exempt: it parses every source, not just the main one.
    private static boolean lacksParseableSource(List<SpecificationSource> sources, OperationProtocol protocol) {
        return protocol != OperationProtocol.GRPC && StringUtils.isBlank(mainSourceText(sources));
    }

    private static void requireProtocol(OperationProtocol protocol) {
        if (protocol == null) {
            throw new SpecificationImportException("Cannot extract schemas: operation protocol is not set.");
        }
    }

    /**
     * Whether the protocol has an extraction path at all. Callers that strip a stored value on the assumption that
     * extraction can rebuild it must gate on this, so the strip and the rebuild stay one decision rather than two
     * protocol lists that can drift apart.
     */
    public static boolean canExtractSchemas(OperationProtocol protocol) {
        return protocol != null && !hasNoSchemaExtraction(protocol);
    }

    // SOAP and METAMODEL carry no request/response schemas by design; degrade to EMPTY_SCHEMAS up front
    // instead of ever attempting a parse.
    private static boolean hasNoSchemaExtraction(OperationProtocol protocol) {
        return protocol == OperationProtocol.SOAP || protocol == OperationProtocol.METAMODEL;
    }

    private static ExtractedSchemas toExtractedSchemas(ParsedOperation operation) {
        return new ExtractedSchemas(
                operation.getSpecification(),
                operation.getRequestSchema(),
                operation.getResponseSchemas());
    }

    private static ParsedOperation matchOperation(List<ParsedOperation> operations, String path, String method) {
        List<ParsedOperation> byPathAndMethod = operations.stream()
                .filter(operation -> Objects.equals(operation.getPath(), path)
                        && method != null
                        && method.equalsIgnoreCase(operation.getMethod()))
                .toList();
        if (byPathAndMethod.size() == 1) {
            return byPathAndMethod.getFirst();
        }
        if (byPathAndMethod.size() > 1) {
            throw new IllegalArgumentException(
                    "Ambiguous operation for path='" + path + "', method='" + method + "'");
        }
        throw new IllegalArgumentException(
                "No operation matched path='" + path + "', method='" + method + "'");
    }
}

File: vscode-extension/src/web/api-services/parsers/OperationSchemaExtractor.ts
import { OpenApiSpecificationParser } from "./OpenApiSpecificationParser";
import { AsyncApiSpecificationParser } from "./AsyncApiSpecificationParser";
import { ProtoSpecificationParser } from "./ProtoSpecificationParser";
import { SpecificationTypeDetector } from "../../services/SpecificationTypeDetector";
import { ASYNC_SPECIFICATION_TYPES } from "../importApiTypes";

export interface ExtractedOperationSchemas {
  specification: Record<string, unknown>;
  requestSchema: Record<string, unknown>;
  responseSchemas: Record<string, unknown>;
}

interface GeneratedOperation {
  name?: string;
  method?: string;
  path?: string;
  specification?: Record<string, unknown>;
  requestSchema?: Record<string, unknown>;
  responseSchemas?: Record<string, unknown>;
}

const emptySchemas = (): ExtractedOperationSchemas => ({
  specification: {},
  requestSchema: {},
  responseSchemas: {},
});

// Extraction only ever matches an already-generated operation by (path, method);
// the id prefix normally built from a specificationId is irrelevant here.
const EXTRACTION_ID = "extract";

// A soft backstop against pathological or corrupted files. The extension parses in
// the extension host, where a runaway parse blocks the editor, so the cap stays here
// even though the backend extractor has none.
const MAX_SOURCE_LENGTH = 5_000_000;

type ExtractableProtocol = "HTTP" | "ASYNC" | "GRPC";

/**
 * Rebuilds one operation's request/response schemas and its specification slice
 * on demand from the raw specification source, mirroring the runtime-catalog
 * backend's `OperationSchemaExtractor`: parse the raw file, match the operation
 * by (path, method), and hand back the same values the full per-protocol parser
 * would have produced at import — reusing each parser's `createOperationsFromX`
 * rather than reimplementing schema production.
 *
 * SOAP and GraphQL carry no schemas by design (mirrors the backend's
 * `hasNoSchemaExtraction`). Any parse failure, oversized source, or unmatched
 * operation degrades gracefully to empty values instead of throwing.
 */
export class OperationSchemaExtractor {
  /**
   * Whether reading a model of this type can rebuild an operation's
   * specification slice from the raw source. The writer strips the stored slice
   * only when this holds, so the strip and the rebuild stay one decision rather
   * than two lists that drift apart — the same gate runtime-catalog's
   * `ServiceSerializer.stripOperationSpecifications` uses.
   *
   * `graphql` is the one place this says no where the backend says yes: the
   * extractor has no graphql path, so a stripped slice would be lost.
   */
  static canRebuildSpecification(specificationType?: string): boolean {
    const type = (specificationType ?? "").toLowerCase();
    return type === "openapi" || type === "asyncapi" || type === "protobuf";
  }

  static async extract(
    rawContent: string | null | undefined,
    formatHint: string | undefined,
    path: string,
    method: string,
  ): Promise<ExtractedOperationSchemas> {
    if (!rawContent || rawContent.length > MAX_SOURCE_LENGTH) {
      return emptySchemas();
    }

    const protocol = this.resolveProtocol(formatHint, rawContent);
    if (!protocol) {
      return emptySchemas();
    }

    try {
      const operations = await this.parseAllOperations(protocol, rawContent);
      const matched = this.matchOperation(operations, path, method);
      return matched
        ? {
            specification: matched.specification ?? {},
            requestSchema: matched.requestSchema ?? {},
            responseSchemas: matched.responseSchemas ?? {},
          }
        : emptySchemas();
    } catch (error) {
      console.warn(
        `[OperationSchemaExtractor] Failed to extract schemas for ${method} ${path}:`,
        error,
      );
      return emptySchemas();
    }
  }

  private static resolveProtocol(
    formatHint: string | undefined,
    rawContent: string,
  ): ExtractableProtocol | null {
    const fromHint = this.protocolFromFormat(formatHint);
    return fromHint !== undefined ? fromHint : this.sniffProtocol(rawContent);
  }

  /**
   * `undefined` means the hint is missing or unrecognized (fall back to
   * sniffing); `null` means it names a protocol that carries no schemas.
   */
  private static protocolFromFormat(
    formatHint: string | undefined,
  ): ExtractableProtocol | null | undefined {
    const upper = (formatHint ?? "").toUpperCase();
    if (upper === "HTTP") {
      return "HTTP";
    }
    if ((ASYNC_SPECIFICATION_TYPES as ReadonlySet<string>).has(upper)) {
      return "ASYNC";
    }
    if (upper === "GRPC") {
      return "GRPC";
    }
    if (upper === "SOAP" || upper === "GRAPHQL") {
      return null;
    }
    return undefined;
  }

  private static sniffProtocol(rawContent: string): ExtractableProtocol | null {
    if (this.isLikelyWsdlContent(rawContent)) {
      return null; // SOAP: no schema extraction by design
    }

    const parsed = SpecificationTypeDetector.parse(rawContent);
    if (parsed && typeof parsed === "object") {
      if (SpecificationTypeDetector.isAsyncApi(parsed)) {
        return "ASYNC";
      }
      if (SpecificationTypeDetector.isOpenApiOrSwagger(parsed)) {
        return "HTTP";
      }
    }

    if (this.isLikelyProtoContent(rawContent)) {
      return "GRPC";
    }

    // GraphQL and anything unrecognized degrade to empty schemas.
    return null;
  }

  private static isLikelyWsdlContent(content: string): boolean {
    const snippet = content.slice(0, 512).toLowerCase();
    return (
      snippet.includes("http://schemas.xmlsoap.org/wsdl") ||
      snippet.includes("http://www.w3.org/ns/wsdl")
    );
  }

  private static isLikelyProtoContent(content: string): boolean {
    return (
      /\bsyntax\s*=\s*["']proto[23]["']/.test(content) ||
      (/\bservice\s+\w+\s*\{/.test(content) && /\brpc\s+\w+\s*\(/.test(content))
    );
  }

  private static async parseAllOperations(
    protocol: ExtractableProtocol,
    rawContent: string,
  ): Promise<GeneratedOperation[]> {
    switch (protocol) {
      case "HTTP": {
        const data =
          await OpenApiSpecificationParser.parseOpenApiContent(rawContent);
        return OpenApiSpecificationParser.createOperationsFromOpenApi(
          data,
          EXTRACTION_ID,
        );
      }
      case "ASYNC": {
        const data =
          await AsyncApiSpecificationParser.parseAsyncApiContent(rawContent);
        return AsyncApiSpecificationParser.createOperationsFromAsyncApi(
          data,
          EXTRACTION_ID,
        );
      }
      case "GRPC": {
        const data =
          await ProtoSpecificationParser.parseProtoContent(rawContent);
        return ProtoSpecificationParser.createOperationsFromProto(
          data,
          EXTRACTION_ID,
        );
      }
    }
  }

  /**
   * Matches by exact (path, method), mirroring the backend's `matchOperation`.
   * Returns undefined (never throws) when the filter is not a single hit, so a
   * miss or an ambiguous match degrades to empty schemas rather than picking a
   * wrong operation.
   */
  private static matchOperation(
    operations: GeneratedOperation[],
    path: string,
    method: string,
  ): GeneratedOperation | undefined {
    const byPathAndMethod = operations.filter(
      (op) =>
        op.path === path &&
        !!method &&
        op.method?.toUpperCase() === method.toUpperCase(),
    );
    return byPathAndMethod.length === 1 ? byPathAndMethod[0] : undefined;
  }
}
